engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/internal/PairwiseDoubleRingBuffer.java

```diff
@@ -0,0 +1,297 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+/*
+ * ---------------------------------------------------------------------------------------------------------------------
+ * AUTO-GENERATED CLASS - DO NOT EDIT MANUALLY - for any changes edit PairwiseFloatRingBuffer and regenerate
+ * ---------------------------------------------------------------------------------------------------------------------
+ */
+package io.deephaven.engine.table.impl.updateby.internal;
+
+import gnu.trove.list.array.TIntArrayList;
+import io.deephaven.chunk.WritableDoubleChunk;
+import io.deephaven.chunk.attributes.Values;
+import io.deephaven.util.SafeCloseable;
+
+import java.util.NoSuchElementException;
+
+/***
+ * Store this data in the form of a binary tree where the latter half of the chunk is treated as a ring buffer and pairwise results of the `DoubleFunction` are stored in the parent nodes.  We do lazy evaluation by maintaining a 'dirty' `BitSet` and compute the ultimate pairwise result only when requested by `evaluate()'
+ *
+ * To keep the parent-node finding math easy and consistent between the ring buffer and the computation tree, the binary tree is shifted by one index so the root (and final result of computation) ends up in index 1 (instead of 0 which is un-used)
+ */
+
+public class PairwiseDoubleRingBuffer implements SafeCloseable {
+    // use a sized double chunk for underlying storage
+    private WritableDoubleChunk<Values> storageChunk;
+//    private final BitSet dirtyBits;
+    private final TIntArrayList dirtyIndices;
+    private boolean allDirty;
+
+    private final DoubleFunction pairwiseFunction;
+    private final double emptyVal;
+
+    // this measures internal storage capacity (chunk is twice this size)
+    private int capacity;
+    private int chunkSize;
+
+    private int head;
+    private int tail;
+
+    @FunctionalInterface
+    public interface DoubleFunction {
```

Comment:
True, but I couldn't find the equivalent for `FloatBinaryOperator` and these are replicated classes.  I didn't want to use `BinaryOperator<T>` because I thought I would be paying boxing/unboxing taxes for each call.